### DISCUSSION TITLE
Integration of heat maps into Remote Usability LAB

### DISCUSSION THREAD

**Discussion by sushant4612** [ruxailab/RUXAILAB#379] [General]
Correct me if I am wrong.

So, I'll delve into depth as I don't have a lot of knowledge regarding heatmaps. Here are some points I've gathered:

## What is heatmaps?
- Heatmaps serve as a form of data representation, similar to charts and graphs. Website heatmaps visualize website user behavior and are used as analytics tools for tracking and predicting users behavior and visual attention flow.

## Types of heatmaps
There are several types of heatmaps, but the idea list mentioned these three: scroll maps, click maps, and move maps.

- A scroll map is a visual representation of your visitors scrolling behavior, allowing you to see how far down the page your visitors have scrolled. It indicates which sections the majority of your users spend time in.

- A click map helps you see if your visitors click where you want them to on your landing pages. Gathering clicking data on CTA buttons and navigation is only one use case of click heatmaps. You can also see clicks on objects that are not, in fact, clickable.

- A mouse tracking heatmap indicates areas over which visitors hover their cursor the most.

## What i understand from the idea

There are two tests available in the remote testing lab: 
- usability user test 
- usability heuristic test
<img width="1466" alt="Screenshot 2024-03-17 at 5 35 04 PM" src="https://github.com/ruxailab/RUXAILAB/assets/109458485/4f8db5a6-17fa-4fc9-876e-bcbc6b4cdaca">

We should create a `usability heatmap test`, which consists of the heatmap test.

##
In addition to the three tests, there are two more heatmap tests available: 
- eye tracking heatmaps 
- AI-generated attention heatmaps

They can also be added, but for eye tracking, i think there is already one project ongoing, and the AI-generated attention heatmap is just the AI version of the eye tracking heatmap.

## Doubt
When I was reading about heatmaps, I discovered some tools like Hotjar, Mouseflow, etc., but they are free for a limited time. So, I think we need to create it ourselves. I really want to know others opinions on it.

**Comment by jvJUCA** [comment]
Hi @sushant4612 ,

Thank you for sharing your insights on heatmaps! It's great to see your interest and willingness to delve into this topic.

Regarding your doubt about using existing tools like Hotjar and Mouseflow versus creating our own solution, it's an important consideration. We've been contemplating a similar approach for implementing the heatmap feature, possibly aligning it with our existing eye tracking project. This could involve exploring options such as developing a browser extension, a desktop application, or even integrating a function directly into the page's console for testing purposes. We're still in the brainstorming phase and evaluating the most effective approach.

Your commitment to exploring different perspectives and solutions is much appreciated. If you have any further thoughts or suggestions as we continue to refine our plans, please feel free to share them.

Thanks again for your dedication and contributions to the discussion!

Best regards,
@jvJUCA

**Comment by sushant4612** [comment]
Hey @jvJUCA , thanks for your response.

Using existing tools has a problem, they are free for a limited time only. I think we should create our own similar tools. It can be challenging, but it could be a better alternative.

Also, I need some help in writing a proposal. Could you please guide me so that I can write a good proposal?

**Reply by jvJUCA** [reply]
Thank you for your response and for sharing your thoughts. I completely understand your concern regarding the limited availability of existing tools. Creating our own similar tools indeed presents challenges but could offer a more sustainable solution in the long run.

Regarding writing the proposal, I'm glad you reached out for guidance. Here's what I suggest:

- Download the proposal template sample from the GSoC repository in our organization.
- Take some time to review good examples of proposals submitted to GSoC in the past. You can find these examples by searching online or through the GSoC website.
- Use the template and examples as a guide to structure and compose your proposal.
- Once your proposal is ready, submit it through the GSoC website to our organization.

If you need any further assistance or have specific questions while writing your proposal, feel free to reach out. I'm here to help!

Best regards,
@jvJUCA

**Reply by sushant4612** [reply]
Thanks, @jvJUCA . I am working on the proposal. I will let you know if I have any doubts.

**Comment by ankitmrj** [comment]
Respected Sir @jvJUCA ,


I am first year undergraduate student, of Computer Science at State University from India. I am thrilled to work on project " Integration of heat maps into Remote Usability LAB ".  I have completed all analysis and extract the knowledge of Heat Maps. As student  of statistics in previous days it easier to gain out and review wide data that help me command over this project.
Heatmapping can also be useful for differentiating between links when there is more than one link on a page that goes to a particular destination page. Some analytics tools, such as Google Analytics, treat links to the same destination as the same link.
There are some Heatmaps tool available for free ,but creating our own tools help to use in our own way .

Please help me to learn and share more idea to this project.
How  to get template for proposal ?.


Thanking You !

Waiting for your positive response.

**Reply by KarinePistili** [reply]
Hello @ankitmrj , the idea behind this project is to create our own analysis that can be used on RUXAILAB for the usability tests. I suggest you to take a look at RUXAILAB main repository, and also right a proposal regarding what you would like to accomplish. You can find proposal template and other gsoc information here: https://github.com/ruxailab/gsoc.

More information on: https://github.com/orgs/ruxailab/discussions/373
Or join our discord: https://discord.gg/4pNs7SvA

**Reply by ankitmrj** [reply]
Thank you @KarinePistili sir for giving valuable response.
I'll follow your above instructions and make better proposal. 
Looking forward to contribute and work together.

Thank you sir from my bottom of heart.

**Reply by KarinePistili** [reply]
Great!

**Comment by Tufailahmed-Bargir** [comment]
Mr @jvJUCA , please resolve my doubts, <br>

The skills to this project is not mentioned, a person knowing JavaScript can successfully take up this project ???
also please tell me who has more potential of being accepted in this project??
what matters more a good proposal or the contributions to the ruxailab???

**Reply by KarinePistili** [reply]
Hello @Tufailahmed-Bargir 

The skills for this would be good knowledge with Javascript, Vue Framework, Git and any other technologiy you suggest on your proposal. We also would like communicative people for all projects.

The evaluation process to select contributors will first analyze all proposals, and rank the best ones. After that we would take a look on contributions and skills.

So I suggest you to focus on your proposal and engage with the community ;)

**Reply by sushant4612** [reply]
Is it correct, @KarinePistili , that the idea is to to add a new test beside the two existing ones?

**Reply by KarinePistili** [reply]
It depends on what idea you are talking about, in the case of the heatmaps (which is the discussion topic), we would like it to be integrated with the User Testing.

**Comment by Z-Jonathan** [comment]
Hi @KarinePistili 

I hope that you're having a great day. I heard that we had the option to submit our proposals for the project earlier to receive feedback, and then refine and resubmit it afterwards before the deadline. Is that correct?

Thank You Again!

**Reply by KarinePistili** [reply]
Google Summer of Code allows editing proposals before deadline on the dashboard. If you need to make modifications after talking to a mentor, I believe it is within the program rules if it is inside de editing deadline.

**Comment by Tufailahmed-Bargir** [comment]
@KarinePistili , Thanks for your valuable response mam, <br>
please don't mind to answer these questions
1) How much pages must a proposal have.
2) The proposal must follow the gsoc temple provided by uraxailab or only the topics which are mentioned in that template
3) How one can reviewed / or submit  the proposal to the mentor for review....

and also please tell us as Mr @jvJUCA , replied <br>
![image](https://github.com/ruxailab/RUXAILAB/assets/142114244/79eafa01-55dc-437a-9a49-7b5270f6137d)
what exactly it is DESTOP APPLICATION, CHROME EXTENTION, or a FUNCTION directly integrated into the code, for our betterment of how we will implement it and how we will spend our time

**Reply by KarinePistili** [reply]
Hello @Tufailahmed-Bargir , we would like to receive the proposal having all the content we asked for. Use as much pages as needed to present the information, but focus on important things, too many pages won't be good as it would be harder to understand what you are going for. 

If you want feedback about your idea you reach out to a mentor on discord, email or via discussion. Tag the mentor or send a private message with your questions.

Regarding the question about the comment, @jvJUCA is listing possibilities of implementation for this specific feature.

How you will implement, and how you will spend your time is up to you to propose inside your proposal document. If you have any question reach out to a mentor to give you feedback.

Hope it answers your questions ;)

**Reply by Tufailahmed-Bargir** [reply]
Thanks mam 😊 , all my doubts are clear now.
Also I have send my proposal in pdf format in discourt please review it and tell me the changes I should make..

**Comment by sushant4612** [comment]
Is creating a library for heatmap or adding an embedding script, which one is the better approach? Both have advantages and disadvantages for instance, an embedding script does not provide control, but creating a library can give users control. I'd like to know @KarinePistili opinion on this

**Reply by KarinePistili** [reply]
We want to create our own, not use an existing one

**Reply by sushant4612** [reply]
Yes, I understand that we need to create our own. But should we do it similar to how other applications are doing it, like providing a script that we embed in the head, or should we consider something else? That's my doubt @KarinePistili

**Reply by KarinePistili** [reply]
Oh ok, I understand now. It would be nice if we can consider doing studies without the need to modify the website code we want to analyse. For example, let's suppose a research team has developed a calendar website, and wants to make a user testing comparing their application with google agenda. We don't have access to google agenda code, so we won't be able to add a script tag to capture the events. For that, we would need a solution that does not need to introduce any code.

**Reply by sushant4612** [reply]
Yes, that's the problem the code, which we don't have, how to add a script to them. So, what is required is that we should pass a URL, and then it should start creating heatmaps and generate output. Actually I'm getting a bit confused about what exactly needs to be created because without changing the website's code, how can we track user interaction with the website? If you have any ideas or resources that I can refer to, it would be helpful for me. @KarinePistili

**Reply by KarinePistili** [reply]
Here on this issue, there some ideas people have been gathering about this: https://github.com/ruxailab/RUXAILAB/issues/253. In fact this is really a challenging one, as we want to find out the type of implementation that attends the most use cases.

**Reply by sushant4612** [reply]
Okay, so I just found out that there's a feature in Hotjar to add a URL and get heatmaps based on it. So, the idea behind it is to render the web page by fetching the HTML content of the specified page and then simulate the page in a virtual environment, just as it looks on the website. After that, they perform mouse clicks and other tests. @KarinePistili

**Reply by KarinePistili** [reply]
Yes, that's it

**Reply by sushant4612** [reply]
I will try my best to write the best proposal accordingly, which will be the best idea. Once I complete it, I will share it with you for review.

**Reply by KarinePistili** [reply]
Great! :smile:

**Reply by sushant4612** [reply]
Hey @KarinePistili , the idea I mentioned to you yesterday is quite challenging because rendering and simulating a full web page can be complex, given modern web technologies such as JavaScript, dynamic content loading, and Single-Page Applications (SPAs). Accurately replicating website behavior in a virtual environment can be resource-intensive and challenging to maintain. Additionally, there are security concerns as websites often implement security measures that may prevent external tools from accessing their content or executing scripts.

**Reply by sushant4612** [reply]
After analyzing the two more approaches for creating a heatmap tool, the first involves developing a browser extension. This extension primarily focuses on capturing user interactions such as clicks and scrolls on webpages. It achieves this by injecting a script into the webpages, enabling the tracking of interactions and sending the data to a server for processing and heatmap generation. The second approach entails utilizing a session replay tool, which captures a wider range of data including user interactions, page loads, time spent on elements, and potentially mouse movements. This tool is designed to record user activity on webpages, possibly including screenshots or visual representations of the user's screen. Additionally, it requires functionalities for storing, managing, and replaying these recordings for analysis.

**Reply by sushant4612** [reply]
Can you provide me with an idea on which approach I should consider for writing my proposal? @KarinePistili

**Reply by sushant4612** [reply]
I searched extensively but couldn't find any ideas on how to implement it without adding any script, as all the approaches under the hood involve adding some code for conducting the user test.

**Reply by sushant4612** [reply]
hey @jvJUCA could you please help me in this

**Reply by KarinePistili** [reply]
Hello @sushant4612 , thanks for the research. I believe you can go with the script option, as it looks more viable to implement and with a more success rate.

**Reply by sushant4612** [reply]
Hey @KarinePistili can you please review my proposal which I sended you on discord
